engine/table/src/main/java/io/deephaven/engine/table/impl/updateby/emstd/BaseBigNumberEmStdOperator.java

```diff
@@ -0,0 +1,248 @@
+        super(pair, affectingColumns, rowRedirection, timestampColumnName, 0, 0, false,
+                BigDecimal.class);
+
+        this.control = control;
+        this.valueSource = valueSource;
+        this.mathContext = mathContext;
+        this.reverseWindowScaleUnits = windowScaleUnits;
+
+        if (sourceRefreshing) {
+            if (rowRedirection != null) {
+                // region create-dense
+                this.maybeEmaInnerSource = new ObjectArraySource<>(BigDecimal.class);
+                // endregion create-dense
+                this.emaSource = WritableRedirectedColumnSource.maybeRedirect(rowRedirection, maybeInnerSource, 0);
+            } else {
+                this.maybeEmaInnerSource = null;
+                // region create-sparse
+                this.emaSource = new ObjectSparseArraySource<>(BigDecimal.class);
+                // endregion create-sparse
+            }
+        } else {
+            this.maybeEmaInnerSource = null;
+            this.emaSource = null;
+        }
+
+        if (timestampColumnName == null) {
+            // tick-based, pre-compute alpha and oneMinusAlpha
+            opAlpha = computeAlpha(-1, reverseWindowScaleUnits);
+            opOneMinusAlpha = computeOneMinusAlpha(opAlpha);
+        } else {
+            // time-based, must compute alpha and oneMinusAlpha for each time delta
+            opAlpha = null;
+            opOneMinusAlpha = null;
+        }
+    }
+
+    @Override
+    public void initializeCumulative(@NotNull final UpdateByOperator.Context updateContext,
+            final long firstUnmodifiedKey,
+            final long firstUnmodifiedTimestamp,
+            @NotNull final RowSet bucketRowSet) {
+        super.initializeCumulative(updateContext, firstUnmodifiedKey, firstUnmodifiedTimestamp, bucketRowSet);
+
+        final Context ctx = (Context) updateContext;
+        // rely on the caller to validate this is a valid timestamp (or NULL_LONG when appropriate)
+        ctx.lastStamp = firstUnmodifiedTimestamp;
+        if (firstUnmodifiedKey != NULL_ROW_KEY) {
+            ctx.curVal = outputSource.get(firstUnmodifiedKey);
+            ctx.curEma = emaSource.get(firstUnmodifiedKey);
+            if (ctx.curVal == null) {
+                ctx.curVariance = BigDecimal.ZERO;
+            } else {
+                ctx.curVariance = ctx.curVal.pow(2);
+            }
+        }
+    }
+
+    @Override
+    public void startTrackingPrev() {
+        super.startTrackingPrev();
+        if (emaSource != null) {
+            emaSource.startTrackingPrevValues();
+            if (rowRedirection != null) {
+                assert maybeEmaInnerSource != null;
+                maybeEmaInnerSource.startTrackingPrevValues();
+            }
+        }
+    }
+
+    // region Shifts
+    @Override
+    public void applyOutputShift(@NotNull final RowSet subRowSetToShift, final long delta) {
+        super.applyOutputShift(subRowSetToShift, delta);
+        if (emaSource != null) {
```

Comment:
Should this instead be: `rowRedirection == null`? I suspect this code path throws a ClassCastException when a rowRedirection is in use.

Reply:
A little bit complicated, but this is correct as is.  `applyOutputShift()` is only called when `rowRedirection == null` because the redirection is managed at a higher layer in `UpdateBy.java`.  Here is the caller, it confirms that the output table is not redirected before calling:
```
                        if (!redirHelper.isRedirected() && upstream.shifted().nonempty()) {
                            // Shift the non-redirected output sources now, after parallelPopulation.
                            upstream.shifted().apply((begin, end, delta) -> {
                                try (final RowSet subRowSet = source.getRowSet().prev().subSetByKeyRange(begin, end)) {
                                    for (UpdateByOperator op : win.getOperators()) {
                                        op.applyOutputShift(subRowSet, delta);
                                    }
                                }
                            });
                        }
```